### PR TITLE
Icon factory cleanups

### PR DESCRIPTION
**Pull request by jennej**
refactors not (yet?) addressed that i’ve noticed while doing this work and have somehow managed not to get terminally distracted by
- the insane amount of code duplication around the resolution independant icons
- the bad code smell in IndirectCurrentAccessibleTypeIcon
- the two competing iconFactoryManagers  + all the other ways we currently choose to manage our factories
- PersonResourceFactory class + GroupIcon + CollageIcon classes who seem to think we’re embarking on a crazy multi-icon mission
- the object/camera marker icons look sad.  are we rendering them twice?  the sceneeditor dropdown of the same icons is much nicer

### CODE REVIEW

File: core/ide/src/main/java/org/alice/ide/instancefactory/croquet/views/icons/IndirectCurrentAccessibleTypeIcon.java [hunk 1]
```diff
@@ -57,6 +57,11 @@
 /**
  * @author Dennis Cosgrove
  */
+
+
+// this used by the object properties tab in the scene editor
+// this is not ideal, since we get a new icon from the factory on Every Paint.
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/instancefactory/croquet/views/icons/IndirectCurrentAccessibleTypeIcon.java:64]: So you are saying these factories pollute?
Do you have any ideas on how to reduce it?

File: core/ide/src/main/java/org/alice/ide/instancefactory/croquet/views/icons/IndirectCurrentAccessibleTypeIcon.java [hunk 2]
```diff
@@ -57,6 +57,11 @@
 /**
  * @author Dennis Cosgrove
  */
+
+
+// this used by the object properties tab in the scene editor
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/instancefactory/croquet/views/icons/IndirectCurrentAccessibleTypeIcon.java]: I think a word was dropped here. Maybe "is"?

File: core/ide/src/main/java/org/alice/stageide/iconfactory/StoryIconFactoryManager.java [hunk 3]
```diff
@@ -54,28 +54,23 @@
...
  */
+
 public class StoryIconFactoryManager implements IconFactoryManager {
+
   public StoryIconFactoryManager() {
   }
 
   @Override
   public IconFactory getIconFactory(UserField field, IconFactory fallbackIconFactory) {
-    final boolean IS_READY_FOR_PRIME_TIME = true;
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/iconfactory/StoryIconFactoryManager.java:63]: I guess it is ready for prime time.

File: core/ide/src/main/java/org/alice/stageide/icons/ColorIconFactory.java [hunk 4]
```diff
@@ -55,6 +55,8 @@
 public class ColorIconFactory extends ResolutionIndependantIconFactory {
   private final Color color;
 
+  protected double defaultAspectRatio = 1.0;
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/icons/ColorIconFactory.java]: Why was this added? It seems to be unused.

File: core/croquet/src/main/java/org/lgna/croquet/icon/AbstractIconFactory.java [hunk 5]
```diff
@@ -93,46 +89,55 @@ public final Icon getIcon(Dimension size) {
...
+  @Override
+  public final Icon getIconToFit(Dimension maxSize) {
+    // getDefaultSize may be overriden, will return null if not.
     Dimension defaultSize = this.getDefaultSize(null);
-    if (defaultSize != null) {
-      return defaultSize.width / (double) defaultSize.height;
+    if (defaultSize == null) {
+      // If there was no default aspect ratio, we can give back the max size
+      return getIconExactSize(maxSize);
     } else {
```
Comment (dabeshouse) [inline core/croquet/src/main/java/org/lgna/croquet/icon/AbstractIconFactory.java:110]: I know it is the only line in the method you didn't change, but this else is redundant.

File: core/ide/src/main/java/org/alice/ide/member/views/MemberTabView.java [hunk 6]
```diff
@@ -80,7 +80,7 @@ public abstract class MemberTabView extends MigPanel {
     AddMethodMenuModel addMethodMenuModel = composite.getAddMethodMenuModel();
     if (addMethodMenuModel != null) {
       this.popupButton = addMethodMenuModel.getPopupPrepModel().createPopupButton();
-      this.popupButton.setClobberIcon(PlusIconFactory.getInstance().getIcon(new Dimension(16, 16)));
+      this.popupButton.setClobberIcon(PlusIconFactory.getInstance().getIconToFit(new Dimension(16, 16)));
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/member/views/MemberTabView.java]: Could use `Theme.EXTRA_SMALL_SQUARE_ICON_SIZE`

File: core/ide/src/main/java/org/alice/stageide/showme/ShowMeHowToAddGalleryModelsIteratingOperation.java [hunk 7]
```diff
@@ -66,7 +66,7 @@ private ShowMeHowToAddGalleryModelsIteratingOperation() {
   @Override
   protected void localize() {
     super.localize();
-    this.setButtonIcon(ShowMeIconFactory.getInstance().getIcon(new Dimension(16, 16)));
+    this.setButtonIcon(ShowMeIconFactory.getInstance().getIconToFit(new Dimension(16, 16)));
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/showme/ShowMeHowToAddGalleryModelsIteratingOperation.java]: Could use `Theme.EXTRA_SMALL_SQUARE_ICON_SIZE`

File: core/ide/src/main/java/org/alice/ide/croquet/components/gallerybrowser/GalleryDragComponent.java [hunk 8]
```diff
@@ -203,7 +203,7 @@ private void addSuperclassIcon(Class<?> modelResourceCls) {
           IconFactory iconFactory = IconFactoryManager.getIconFactoryForResourceCls((Class<ModelResource>) modelResourceInterface);
           if (iconFactory != null) {
             final Dimension SUPER_CLASS_ICON_SIZE = new Dimension(32, 24);
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/croquet/components/gallerybrowser/GalleryDragComponent.java]: `Theme.EXTRA_SMALL_RECT_ICON_SIZE`?

File: core/ide/src/main/java/org/alice/ide/croquet/models/cascade/SimpleExpressionFillIn.java [hunk 9]
```diff
@@ -95,9 +95,9 @@ protected Icon getLeadingIcon(ItemNode<? super E, Void> step) {
               Dimension size = new Dimension(24, 18);
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/croquet/models/cascade/SimpleExpressionFillIn.java]: `Theme.EXTRA_SMALL_RECT_ICON_SIZE`?

File: core/ide/src/main/java/org/alice/ide/Theme.java [hunk 10]
```diff
@@ -54,9 +54,14 @@
  * @author Dennis Cosgrove
  */
 public interface Theme {
-  public static final Dimension DEFAULT_SMALLER_ICON_SIZE = new Dimension(32, 24);
-  public static final Dimension DEFAULT_SMALL_ICON_SIZE = new Dimension(40, 30);
-  public static final Dimension DEFAULT_LARGE_ICON_SIZE = new Dimension(120, 90);
+  public static final Dimension EXTRA_SMALL_RECT_ICON_SIZE = new Dimension(32, 24);
+  public static final Dimension SMALL_RECT_ICON_SIZE = new Dimension(40, 30);
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/ide/Theme.java]: The size descriptors were already here, but with the new sizes added this small and extra small look closer to medium and small in square. Consider renaming these two up to small and medium.
Could go another step and add a new extra small at the 24x16 to be used in ResourceNodeTreeState

File: core/ide/src/main/java/org/alice/stageide/StageIDE.java [hunk 11]
```diff
@@ -235,7 +235,7 @@ private Icon getIconFor(AbstractField field) {
           JavaField javaField = (JavaField) field;
           org.lgna.story.Color color = (org.lgna.story.Color) ReflectionUtilities.get(javaField.getFieldReflectionProxy().getReification(), null);
           Color awtColor = EmployeesOnly.getAwtColor(color);
-          return new ColorIconFactory(awtColor).getIcon(new Dimension(15, 15));
+          return new ColorIconFactory(awtColor).getIconToFit(new Dimension(15, 15));
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/StageIDE.java]: We didn't have enough different sizes and couldn't use 16x16. Oy.

File: core/ide/src/main/java/org/alice/stageide/modelresource/ResourceNodeTreeState.java [hunk 12]
```diff
@@ -63,7 +63,7 @@
 public class ResourceNodeTreeState extends CustomSingleSelectTreeState<ResourceNode> {
 
   //todo
-  private static final Icon EMPTY_ICON = new EmptyIcon(0, Theme.DEFAULT_SMALL_ICON_SIZE.height);
+  private static final Icon EMPTY_ICON = new EmptyIcon(0, Theme.SMALL_RECT_ICON_SIZE.height);
 
   private static final Dimension BUTTON_ICON_SIZE = new Dimension(24, 18);
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/modelresource/ResourceNodeTreeState.java]: Could be pushed into Theme.

File: core/ide/src/main/java/org/alice/stageide/properties/MutableRiderVehicleAdapter.java [hunk 13]
```diff
@@ -121,7 +121,7 @@ public static Icon getIconForVehicle(SThing vehicle) {
       UserField field = IDE.getActiveInstance().getSceneEditor().getFieldForInstanceInJavaVM(vehicle);
       if (field != null) {
         IconFactory iconFactory = IconFactoryManager.getIconFactoryForField(field);
-        return iconFactory.getIcon(new Dimension(24, 18));
+        return iconFactory.getIconToFit(new Dimension(24, 18));
```
Comment (dabeshouse) [inline core/ide/src/main/java/org/alice/stageide/properties/MutableRiderVehicleAdapter.java]: Same value as `ResourceNodeTreeState` that could be in `Theme`.

**Review by dabeshouse — CHANGES_REQUESTED** [review]
All good changes.
Happy for each removal of `// pass`.
Only minor suggestions for you to consider.

File: core/ide/src/main/java/org/alice/stageide/StageIDE.java (hunk 11 above)
Comment (jennej) [reply core/ide/src/main/java/org/alice/stageide/StageIDE.java]: this actually looks fine as an XS square, so i'm improving all of our lives

File: core/ide/src/main/java/org/alice/stageide/iconfactory/StoryIconFactoryManager.java (hunk 3 above)
Comment (jennej) [reply core/ide/src/main/java/org/alice/stageide/iconfactory/StoryIconFactoryManager.java:63]: the prime is now

**Review by dabeshouse — APPROVED** [review]
